integration/spark/src/main/spark3.1/org/apache/spark/sql/SparkVersionAdapter.scala

```diff
@@ -0,0 +1,549 @@
+    }
+    tableInfo.setTransactionalTable(isTransactionalTable)
+    selectQuery match {
+      case query@Some(q) =>
+        CarbonCreateTableAsSelectCommand(
+          tableInfo = tableInfo,
+          query = query.get,
+          ifNotExistsSet = ifNotExists,
+          tableLocation = tablePath)
+      case _ =>
+        CarbonCreateTableCommand(
+          tableInfo = tableInfo,
+          ifNotExistsSet = ifNotExists,
+          tableLocation = tablePath,
+          external)
+    }
+  }
+
+  def getField(parser: CarbonSpark2SqlParser,
+               schema: Seq[QualifiedColType], isExternal: Boolean = false): Seq[Field] = {
+    schema.map { col =>
+      parser.getFields(col.comment, col.name.head, col.dataType, isExternal)
+    }
+  }
+
+  def supportsBatchOrColumnar(scan: CarbonDataSourceScan): Boolean = {
+    scan.supportsColumnar
+  }
+
+  def createDataset(sparkSession: SparkSession, qe: QueryExecution) : Dataset[Row] = {
+    new Dataset[Row](qe, RowEncoder(qe.analyzed.schema))
+  }
+
+  def createSharedState(sparkContext: SparkContext) : SharedState = {
+    new SharedState(sparkContext, Map.empty[String, String])
+  }
+
+  def translateFilter(dataFilters: Seq[Expression]) : Seq[Filter] = {
+    dataFilters.flatMap(DataSourceStrategy.translateFilter(_,
+      supportNestedPredicatePushdown = false))
+  }
+
+  def getCarbonOptimizer(session: SparkSession,
+                         sessionState: SessionState): CarbonOptimizer = {
+    new CarbonOptimizer(session, sessionState.optimizer)
+  }
+
+  def isCharType(dataType: DataType): Boolean = {
+    dataType.isInstanceOf[CharType]
+  }
+
+  def isVarCharType(dataType: DataType): Boolean = {
+    dataType.isInstanceOf[VarcharType]
+  }
+
+  def getTypeName(s: AbstractDataType): String = {
+    s.defaultConcreteType.typeName
+  }
+
+  def getInsertIntoCommand(table: LogicalPlan,
+                           partition: Map[String, Option[String]],
+                           query: LogicalPlan,
+                           overwrite: Boolean,
+                           ifPartitionNotExists: Boolean): InsertIntoStatement = {
+    InsertIntoStatement(
+      table,
+      partition,
+      Nil,
+      query,
+      overwrite,
+      ifPartitionNotExists)
+  }
+}
+
+case class CarbonBuildSide(buildSide: BuildSide) {
+  def isRight: Boolean = buildSide.isInstanceOf[BuildRight.type]
+  def isLeft: Boolean = buildSide.isInstanceOf[BuildLeft.type]
+}
+
+abstract class CarbonTakeOrderedAndProjectExecHelper(sortOrder: Seq[SortOrder],
+    limit: Int, skipMapOrder: Boolean, readFromHead: Boolean) extends UnaryExecNode {
+  override def simpleString(maxFields: Int): String = {
+    val orderByString = sortOrder.mkString("[", ",", "]")
+    val outputString = output.mkString("[", ",", "]")
+
+    s"CarbonTakeOrderedAndProjectExec(limit=$limit, orderBy=$orderByString, " +
+      s"skipMapOrder=$skipMapOrder, readFromHead=$readFromHead, output=$outputString)"
+  }
+}
```

Comment:
add a new line after class, all these should be caught in CI, please check once

Reply:
done